fix(core): deferred blocks not removing content immediately when animations are enabled

crisbeto opened the pull request:
Fixes an issue where if animations are enabled, deferred blocks don't remove their placeholder blocks immediately from the DOM. The problem is that we register the event handlers in `afterRender` which runs outside the zone, but the logic that removes the DOM nodes during animations is tied to change detection.

These changes resolve the issue by binding the listeners inside the zone. This was the intention from the beginning, I just forgot that `afterRender` runs outside the zone.

Fixes #51970.

crisbeto commented on packages/core/src/render3/instructions/defer_events.ts:
```diff
@@ -59,9 +61,13 @@ export function onInteraction(trigger: Element, callback: VoidFunction) {
-    }
+    // Ensure that the handler runs in the NgZone since it gets
+    // registered in `afterRender` which runs outside.
+    injector.get(NgZone).run(() => {
```
Note that I was having a hard time capturing this behavior in a unit test, because the testing APIs would trigger change detection which would flush the animations.

AndrewKushnir replied:
Thanks for the comment, I haven't seen it while leaving a review comment (https://github.com/angular/angular/pull/51971#pullrequestreview-1651957116). May be we can have a test to make sure that we run event handlers inside NgZone?

crisbeto replied:
@AndrewKushnir I left a note here about why I think it'll be tricky to capture this in a test. We may be able to potentially do it in an e2e test, but we don't have any of those for `@defer` yet.

crisbeto replied:
Hmm, I don't know if we can observe the internal event handlers without introducing another DI token. I ended up adding a dev mode assertion that it's in the `NgZone` which should give us the same result.

AndrewKushnir replied:
> I ended up adding a dev mode assertion that it's in the NgZone which should give us the same result.

That's a great idea!

crisbeto replied:
It looks like the assertion isn't super reliable since Zone.js doesn't seem to patch `addEventListener` in the same way in Node and in the browser so we were getting different test results. I reworked it to spy on `addEventListener` and check that it's called within the zone instead.

AndrewKushnir commented on packages/core/src/render3/instructions/defer_events.ts:
```diff
@@ -59,9 +61,13 @@ export function onInteraction(trigger: Element, callback: VoidFunction) {
+    // registered in `afterRender` which runs outside.
+    injector.get(NgZone).run(() => {
+      for (const name of interactionEventNames) {
+        trigger.addEventListener(name, entry!.listener, eventListenerOptions);
```
In a context of https://github.com/angular/angular/issues/51970 and the animation renderer, should we also use `renderer.listen` here instead (since it contains some extra handling for animations)?

crisbeto replied:
I'm not sure that I follow the note about #51970. I also don't think that we need to go through the renderer here. It wouldn't have prevented this issue, because even if the animation finished, I think that we would've hit some other change detection issue further down the line since things tend to behave weirdly if view get created outside the zone. We've seen this on Components in the past as well.

AndrewKushnir replied:
Sounds good, let's cover this use-case once we have e2e tests added (I've captured this in internal TODOs doc) 👍

AndrewKushnir reviewed (commented):
@crisbeto thanks for the fix 👍

Just left a comment and I was also wondering if we can add a new test (or modify one of the existing one) to reproduce an original issue with animations?

AndrewKushnir commented on packages/core/src/render3/instructions/defer.ts:
```diff
@@ -758,6 +758,7 @@ function triggerDeferBlock(lView: LView, tNode: TNode) {
   const lContainer = lView[tNode.index];
   const injector = lView[INJECTOR]!;
   ngDevMode && assertLContainer(lContainer);
+  ngDevMode && NgZone.assertInAngularZone();
```
👍

crisbeto committed:
fix(core): deferred blocks not removing content immediately when animations are enabled

Fixes an issue where if animations are enabled, deferred blocks don't remove their placeholder blocks immediately from the DOM. The problem is that we register the event handlers in `afterRender` which runs outside the zone, but the logic that removes the DOM nodes during animations is tied to change detection.

These changes resolve the issue by binding the listeners inside the zone. This was the intention from the beginning, I just forgot that `afterRender` runs outside the zone.

Fixes #51970.

crisbeto commented on packages/core/test/acceptance/defer_spec.ts:
```diff
@@ -1227,6 +1227,7 @@ describe('@defer', () => {
         template: 'Primary block content.',
       })
       class NestedCmp {
+        @Input() block!: string;
```
This is unrelated, I just noticed that it was logging a warning.

alxhub commented:
This PR was merged into the repository by commit 4f69d620d94663592780b2875acbc2b1918775f9.
